### DISCUSSION TITLE
Jupyter Notebook installation failed because jsonschema-specifications installation failed

### DISCUSSION THREAD

**Discussion by LucieSauvajon** [python-poetry/poetry#9674] [Q&A] [answered]
Hi, I installed poetry on Windows and I created a project with `poetry init`.
I would like to add notebook package to use jupyter notebook with 'poetry add --dev notebook',
but the installation failed since `Installing jsonschema-specifications (2023.12.1): Failed` :

![image](https://github.com/user-attachments/assets/f851561d-f403-4725-8e20-17c3d3e6ed3c)

Indeed, poetry can't find the file format-annotation of the package jsonschema-specifications.
I check in the jsonschema-specifications files and I found only the following files but no file named "format-annotation" :

![image](https://github.com/user-attachments/assets/8ae7d4db-5429-424f-92dc-1c07e76ed93e)

Do you know how I could fix the installation of Jupyter Notebook through poetry ? (I tried to install jupyter lab but same issue)

**Comment by LucieSauvajon** [comment]
**Issue solved :**

Initially, I tried uninstalling both Poetry and Pipx. Then, I reinstalled Pipx using Pip, followed by reinstalling Poetry.

However, a new issue arose: I had no permission to add packages through the VSCode terminal. I discovered that the folder pypoetry or Cache, which should be located in AppData\Local, had not been created. This indicated an error related to the cache directory.

To resolve this, I changed the cache directory to a new location with proper write permissions by running the following command:
`poetry config cache-dir C:\Users\sauva\.poetry_cache
`

After changing the cache directory, I successfully tested package installations through the VSCode terminal, including the notebook package.


**Comment:**
Prior to this, I had also changed the virtual environment path to:
`poetry config virtualenvs.path C:\Users\sauva\.poetry_venvs
`
However, I am unsure if this change had any direct impact on resolving the issue.

**Comment by LucieSauvajon** [comment]
More details about the solution by chatgpt: 

### **1. Why did your solution work?**

- Permission issues: When Poetry installs packages, it uses two essential directories:
   - The cache directory to store downloaded files from PyPI (like packages).
   - The virtual environment directory to isolate dependencies for each project.

   In your case, the directory C:\Users\sauva\AppData\Local\pypoetry\Cache, which is the default cache directory, didn’t have the necessary permissions for Poetry to write and create cache files.


- **Changing the cache directory:** By configuring Poetry to use a new cache directory (C:\Users\sauva\.poetry_cache) in a location where you have write permissions, you bypassed the permission issues. This allowed Poetry to store cache files without any problems, enabling package installations as expected.


- **Changing the virtual environment path:** By changing the virtual environment path (C:\Users\sauva\.poetry_venvs), you ensured that Poetry creates these environments in a folder under your control (outside of AppData, which can sometimes have permission restrictions). While this may not be directly related to the initial issue, it’s a good practice to keep your virtual environments in a dedicated folder you can easily manage.



### **2. What do these directories represent in Poetry?**
- **Cache directory:** The cache directory in Poetry is where it temporarily stores downloaded files from package repositories like PyPI. This includes tar.gz, wheels, and other files needed to install packages. The cache helps:
   - Reduce download time when installing the same package multiple times (by using files already in the cache).
   - Minimize reliance on internet access for repeated installations.

   By default, this cache is located at C:\Users\sauva\AppData\Local\pypoetry\Cache, but you can move it to a more suitable location using the poetry config cache-dir command.

- **Virtual environment path:** Poetry uses virtual environments to isolate the dependencies of a specific project. This prevents conflicts between different package versions across projects. A virtual environment is a directory where Python and all necessary project packages are installed.

By default, Poetry creates these virtual environments in a cache folder. By setting virtualenvs.path to C:\Users\sauva\.poetry_venvs, you specified a particular location where each project will have its own subdirectory with its virtual environment.

- **Benefits of using a dedicated location:**
   - Easier to manage and clean up since all your virtual environments are in one place.
   - Fewer permission or restriction issues when creating these environments in your user directory, compared to system-restricted locations like AppData.


### **Conclusion**
- **The cache** stores temporary files necessary for installing packages, and by changing this directory to a location with proper permissions, you resolved the permission issues.
- **Virtual environments** isolate each project’s dependencies, and changing their path allows you to better organize and control where these environments are created.

By modifying these paths, you gained more control over the files that Poetry creates, while avoiding permission and system restriction issues.

**Reply by UtpalMattoo** [reply]
![image](https://github.com/user-attachments/assets/9ff17240-17f5-4446-9c5d-38431e805c7a)
A similar problem here...

**Reply by abn** [reply]
@UtpalMattoo this is a closed discussion.

The issue reported here is likely a result of using Python installed from Microsoft Store which uses a sandboxed APP_DATA directory. 

You can work around this issue by either using a Python version installed using an installer from https://python.org or setting cache directory explicitly like indicated at https://github.com/python-poetry/install.python-poetry.org/issues/135#issuecomment-2494033993.

**Comment by abn** [comment, marked as answer]
The issue reported here is likely a result of using Python installed from Microsoft Store which uses a sandboxed APP_DATA directory. 

You can work around this issue by either using a Python version installed using an installer from https://python.org or setting cache directory explicitly like indicated at https://github.com/python-poetry/install.python-poetry.org/issues/135#issuecomment-2494033993.

```sh
poetry config cache-dir C:\Users\username\pypoetry
```
